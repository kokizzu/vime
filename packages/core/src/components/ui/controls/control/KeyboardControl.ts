export interface KeyboardControl {
  /**
   * A pipe (`|`) seperated string of JS key codes, that when caught in a `keydown` event, will
   * trigger a `click` event on the control.
   *
   * @see https://keycode.info
   */
  keyCodes?: string;

  /**
   * If the `keyCodes` prop is provided, this prop can provide a hint to the user inside the
   * tooltip for what keys can be pressed to trigger the control.
   */
  keyboardHint?: string;
}
