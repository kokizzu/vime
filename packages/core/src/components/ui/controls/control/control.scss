vime-control {
  /**
    * @prop --control-border-radius: The border radius of the control.
    * @prop --control-bg: The background of the control.
    * @prop --control-scale: The amount to scale the control up/down by.
    * @prop --control-border: The border of the control.
    * @prop --control-padding: The padding inside the control.
    * @prop --control-color: The text color of the control.
    * @prop --control-focus-bg: The background colour of a control when it is being hovered on or 
    * focused.
    * @prop --control-focus-color: The text colour of a control when it is being hovered on or 
    * focused.
    * @prop --control-tap-highlight: The highlight color when a control is tapped.
    */

  &.hidden {
    display: none;
  }

  & > button {
    display: flex;
    align-items: center;
    flex-direction: row;
    border: var(--control-border);
    cursor: pointer;
    flex-shrink: 0;
    color: var(--control-color);
    background: var(--control-bg, transparent);
    border-radius: var(--control-border-radius);
    padding: var(--control-padding);
    position: relative;
    pointer-events: auto;
    transition: all 0.3s ease;
    transform: scale(var(--control-scale, 1));
    
    &:focus {
      outline: 0;
    }
    
    &.tapHighlight {
      background: var(--control-tap-highlight);
    }

    &.notTouch {
      &:focus,
      &:hover,
      &[aria-expanded='true'] {
        background: var(--control-focus-bg);
        color: var(--control-focus-color);
      }
    }
  }
}