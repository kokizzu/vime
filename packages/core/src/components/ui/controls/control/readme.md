# vime-control

A generic player control that is designed to work with both touch and mouse devices. It also
seamlessly works with `vime-tooltip`, which can be passed in via the default `slot`.

## Visual

<img
  src="https://raw.githubusercontent.com/vime-js/vime/master/packages/core/src/components/ui/controls/control/control.png"
  alt="Vime control component"
/>

<!-- Auto Generated Below -->


## Usage

### Angular

```html title="playback-control.html"
<vime-control
  keys="k"
  [label]="i18n.playback"
  [pressed]="paused"
  (click)="onClick()"
>
  <vime-icon [href]="icon"></vime-icon>
  <vime-tooltip>{{tooltip}} (k)</vime-tooltip>
</vime-control>
```

```ts title="playback-control.ts"
import { Component, ElementRef } from '@angular/core';
import { VimeComponent } from '@vime/angular';

@Component({
  selector: 'playback-control',
  templateUrl: './playback-control.html',
})
class PlaybackControl extends VimeComponent {
  paused = true;

  i18n = {};

  constructor(protected ref: ElementRef) {
    super(['paused', 'i18n']);
  }

  get icon() {
    return this.paused ? '#vime-play' : '#vime-pause';
  }

  get tooltip() {
    return this.paused ? this.i18n.play : this.i18n.pause;
  }

  onClick() {
    this.paused = !this.paused;
  }
}
```


### Html

```html {6-9}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-controls full-width>
      <vime-control label="Playback" keys="k" pressed="false">
        <vime-icon href="#vime-play"></vime-icon>
        <vime-tooltip>Play (k)</vime-tooltip>
      </vime-control>
    </vime-controls>
  </vime-ui>
</vime-player>
```


### React

```tsx {3,20-29}
import React, { useMemo, useRef } from 'react';
import {
  VimeControl,
  VimeIcon,
  VimeTooltip,
  usePlayerContext,
} from '@vime/react';

function PlaybackControl() {
  const ref = useRef(null);
  const [paused, setPaused] = usePlayerContext(ref, 'paused', true);
  const [i18n] = usePlayerContext(ref, 'i18n', {});
  const icon = useMemo(() => (paused ? '#vime-play' : '#vime-pause'), [paused]);
  const tooltip = useMemo(() => (paused ? 'Play' : 'Pause'), [paused]);
  const onClick = () => {
    setPaused(false);
  };

  return render(
    <VimeControl
      keys="k"
      ref={ref}
      label={i18n.playback}
      pressed={paused}
      onClick={onClick}
    >
      <VimeIcon href={icon} />
      <VimeTooltip>{tooltip} (k)</VimeTooltip>
    </VimeControl>
  );
}
```


### Svelte

```tsx
<VimeControl
  keys="k"
  label={$i18n.playback}
  pressed={paused}
  on:click={onClick}
  bind:this={ref}
>
  <VimeIcon href={icon} />
  <VimeTooltip>{tooltip} (k)</VimeTooltip>
</VimeControl>
```

```html {4}
<script lang="ts">
  import {
    usePlayerStore,
    VimeControl,
    VimeIcon,
    VimeTooltip,
  } from '@vime/svelte';

  let ref: VimeControl;

  const { paused, i18n } = usePlayerStore(() => ref);

  const onClick = () => {
    $paused = !$paused;
  };

  $: icon = $paused ? '#vime-play' : '#vime-pause';
  $: tooltip = $paused ? $i18n.play : $i18n.pause;
</script>
```


### Vue

```html {2-10,17,29} title="playback-control.vue"
<template>
  <VimeControl
    keys="k"
    :label="i18n.playback"
    :pressed="paused"
    @click="onClick"
  >
    <VimeIcon :href="icon" />
    <VimeTooltip>{{tooltip}} (k)</VimeTooltip>
  </VimeControl>
</template>

<script>
  import {
    VimeMixin,
    VimeControl,
    VimeIcon,
    VimeTooltip,
  } from "@vime/vue";

  export default {
    mixins: [VimeMixin(['paused', 'i18n'])]

    components: {
      VimeControl,
      VimeIcon,
      VimeTooltip,
    },

    data: {
      paused: true,
      i18n: {},
    },

    computed: {
      icon() {
        return this.paused ? '#vime-play' : '#vime-pause';
      },
      tooltip() {
        return this.paused ? this.i18n.play : this.i18n.pause;
      },
    },

    methods: {
      onClick() {
        this.paused = !paused;
      },
    },
  };,
</script>
```



## Properties

| Property             | Attribute    | Description                                                                                                                                                   | Type                   | Default     |
| -------------------- | ------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------- | ----------- |
| `expanded`           | `expanded`   | If the control has a popup menu, this indicates whether the menu is open or not. Sets the `aria-expanded` property.                                           | `boolean \| undefined` | `undefined` |
| `hidden`             | `hidden`     | Whether the control should be displayed or not.                                                                                                               | `boolean`              | `false`     |
| `identifier`         | `identifier` | The `id` attribute of the control.                                                                                                                            | `string \| undefined`  | `undefined` |
| `keys`               | `keys`       | A slash (`/`) separated string of JS keyboard keys (`KeyboardEvent.key`), that when caught in a `keydown` event, will trigger a `click` event on the control. | `string \| undefined`  | `undefined` |
| `label` _(required)_ | `label`      | The `aria-label` property of the control.                                                                                                                     | `string`               | `undefined` |
| `menu`               | `menu`       | If the control has a popup menu, then this should be the `id` of said menu. Sets the `aria-controls` property.                                                | `string \| undefined`  | `undefined` |
| `pressed`            | `pressed`    | If the control is a toggle, this indicated whether the control is in a "pressed" state or not. Sets the `aria-pressed` property.                              | `boolean \| undefined` | `undefined` |
| `scale`              | `scale`      | Scale the size of the control up/down by the amount given.                                                                                                    | `number`               | `1`         |


## Events

| Event                | Description                                                                                    | Type                   |
| -------------------- | ---------------------------------------------------------------------------------------------- | ---------------------- |
| `vInteractionChange` | Emitted when the user is interacting with the control by focusing, touching or hovering on it. | `CustomEvent<boolean>` |


## Slots

| Slot | Description                                                     |
| ---- | --------------------------------------------------------------- |
|      | Used to pass in the content of the control (text/icon/tooltip). |


## CSS Custom Properties

| Name                      | Description                                                                |
| ------------------------- | -------------------------------------------------------------------------- |
| `--control-bg`            | The background of the control.                                             |
| `--control-border`        | The border of the control.                                                 |
| `--control-border-radius` | The border radius of the control.                                          |
| `--control-color`         | The text color of the control.                                             |
| `--control-focus-bg`      | The background colour of a control when it is being hovered on or focused. |
| `--control-focus-color`   | The text colour of a control when it is being hovered on or focused.       |
| `--control-padding`       | The padding inside the control.                                            |
| `--control-tap-highlight` | The highlight color when a control is tapped.                              |


## Dependencies

### Used by

 - [vime-caption-control](../caption-control)
 - [vime-fullscreen-control](../fullscreen-control)
 - [vime-mute-control](../mute-control)
 - [vime-pip-control](../pip-control)
 - [vime-playback-control](../playback-control)
 - [vime-settings-control](../settings-control)

### Graph
```mermaid
graph TD;
  vime-caption-control --> vime-control
  vime-fullscreen-control --> vime-control
  vime-mute-control --> vime-control
  vime-pip-control --> vime-control
  vime-playback-control --> vime-control
  vime-settings-control --> vime-control
  style vime-control fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
