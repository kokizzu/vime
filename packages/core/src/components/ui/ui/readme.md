# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the 
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop 
is `true`, and the player is not in fullscreen mode.

## Example

```html
<vime-player>
  <!-- ... -->

  <vime-ui>
    <vime-poster />
    <!-- ... -->
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


## Dependencies

### Used by

 - [vime-default-ui](../default-ui)

### Graph
```mermaid
graph TD;
  vime-default-ui --> vime-ui
  style vime-ui fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
