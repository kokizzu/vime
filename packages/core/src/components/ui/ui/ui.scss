vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}