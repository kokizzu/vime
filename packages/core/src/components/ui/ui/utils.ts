import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
