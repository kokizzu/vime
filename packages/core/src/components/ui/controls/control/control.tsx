import {
  h, Host, Component, Prop, State, Element,
  Event, EventEmitter,
  Watch,
} from '@stencil/core';
import { openPlayerWormhole } from '../../../core/player/PlayerWormhole';
import { PlayerProp, PlayerProps } from '../../../core/player/PlayerProp';
import { isNull, isUndefined } from '../../../../utils/unit';
import { Disposal } from '../../../core/player/Disposal';
import { listen } from '../../../../utils/dom';
import { KeyboardControl } from './KeyboardControl';
import { findRootPlayer } from '../../../core/player/utils';

/**
 * @slot - Used to pass in the content of the control (text/icon/tooltip).
 */
@Component({
  tag: 'vime-control',
  styleUrl: 'control.scss',
})
export class Control implements KeyboardControl {
  private button!: HTMLButtonElement;

  private keyboardDisposal = new Disposal();

  @Element() el!: HTMLVimeControlElement;

  @State() describedBy?: string;

  @State() showTapHighlight = false;

  /**
   * @inheritdoc
   */
  @Prop() keys?: string;

  @Watch('keys')
  onKeysChange() {
    this.keyboardDisposal.empty();
    if (isUndefined(this.keys)) return;

    const player = findRootPlayer(this);
    const codes = (this.keys! as string).split('/');

    this.keyboardDisposal.add(listen(player, 'keydown', (event: KeyboardEvent) => {
      if (codes.includes(event.key)) { this.button.click(); }
    }));
  }

  /**
   * The `id` attribute of the control.
   */
  @Prop() identifier?: string;

  /**
   * Whether the control should be displayed or not.
   */
  @Prop() hidden = false;

  /**
   * The `aria-label` property of the control.
   */
  @Prop() label!: string;

  /**
   * If the control has a popup menu, then this should be the `id` of said menu. Sets the
   * `aria-controls` property.
   */
  @Prop() menu?: string;

  /**
   * If the control has a popup menu, this indicates whether the menu is open or not. Sets the
   * `aria-expanded` property.
   */
  @Prop() expanded?: boolean;

  /**
   * If the control is a toggle, this indicated whether the control is in a "pressed" state or not.
   * Sets the `aria-pressed` property.
   */
  @Prop() pressed?: boolean;

  /**
   * Scale the size of the control up/down by the amount given.
   */
  @Prop() scale = 1;

  /**
   * @internal
   */
  @Prop() isTouch: PlayerProps[PlayerProp.IsTouch] = false;

  /**
   * Emitted when the user is interacting with the control by focusing, touching or hovering on it.
   */
  @Event() vInteractionChange!: EventEmitter<boolean>;

  componentWillLoad() {
    this.findTooltip();
    this.onKeysChange();
  }

  disconnectedCallback() {
    this.keyboardDisposal.empty();
  }

  private onTouchStart() {
    this.showTapHighlight = true;
    setTimeout(() => { this.showTapHighlight = false; }, 100);
  }

  private findTooltip() {
    const tooltip = this.el.querySelector('vime-tooltip');
    if (!isNull(tooltip)) this.describedBy = tooltip!.id;
    return tooltip;
  }

  private onShowTooltip() {
    const tooltip = this.findTooltip();
    if (!isNull(tooltip)) tooltip!.active = true;
    this.vInteractionChange.emit(true);
  }

  private onHideTooltip() {
    const tooltip = this.findTooltip();
    if (!isNull(tooltip)) tooltip!.active = false;
    this.button.blur();
    this.vInteractionChange.emit(false);
  }

  private onFocus() {
    this.el.dispatchEvent(new window.Event('focus', { bubbles: true }));
    this.onShowTooltip();
  }

  private onBlur() {
    this.el.dispatchEvent(new window.Event('blur', { bubbles: true }));
    this.onHideTooltip();
  }

  private onMouseEnter() {
    this.onShowTooltip();
  }

  private onMouseLeave() {
    this.onHideTooltip();
  }

  render() {
    return (
      <Host
        class={{
          hidden: this.hidden,
        }}
      >
        <button
          style={{
            transform: `scale(${this.scale})`,
          }}
          class={{
            notTouch: !this.isTouch,
            tapHighlight: this.showTapHighlight,
          }}
          id={this.identifier}
          type="button"
          aria-label={this.label}
          aria-haspopup={!isUndefined(this.menu) ? 'true' : undefined}
          aria-controls={this.menu}
          // eslint-disable-next-line no-nested-ternary
          aria-expanded={!isUndefined(this.menu) ? (this.expanded ? 'true' : 'false') : undefined}
          // eslint-disable-next-line no-nested-ternary
          aria-pressed={!isUndefined(this.pressed) ? (this.pressed ? 'true' : 'false') : undefined}
          aria-hidden={this.hidden ? 'true' : 'false'}
          aria-describedby={this.describedBy}
          onTouchStart={this.onTouchStart.bind(this)}
          onFocus={this.onFocus.bind(this)}
          onBlur={this.onBlur.bind(this)}
          onMouseEnter={this.onMouseEnter.bind(this)}
          onMouseLeave={this.onMouseLeave.bind(this)}
          ref={(el: any) => { this.button = el; }}
        >
          <slot />
        </button>
      </Host>
    );
  }
}

openPlayerWormhole(Control, [
  PlayerProp.IsTouch,
]);
