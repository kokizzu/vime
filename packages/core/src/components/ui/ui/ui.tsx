import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { withPlayerContext } from '../../core/player/PlayerContext';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

/**
 * @slot - Used to pass in UI components for the player.
 */
@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.isVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.isFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        <div>
          {canShowCustomUI && <slot />}
        </div>
      </Host>
    );
  }
}

withPlayerContext(UI, [
  PlayerProp.isVideoView,
  PlayerProp.playsinline,
  PlayerProp.isFullscreenActive,
]);
