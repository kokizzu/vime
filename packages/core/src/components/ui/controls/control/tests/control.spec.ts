import { SpecPage } from '@stencil/core/testing';
import { newUISpecPage } from '../../../ui/tests';
import { Control } from '../control';

let page: SpecPage;
let control: HTMLVimeControlElement;

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [Control],
    '<vime-control>Control</vime-control>',
  ));

  control = page.root!.querySelector('vime-control')!;
});

it('should be structurally sound', async () => {
  expect(control).toMatchSnapshot();
});
