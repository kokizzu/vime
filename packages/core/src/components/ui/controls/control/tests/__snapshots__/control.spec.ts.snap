// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-control>
  <!---->
  <button aria-haspopup="false" aria-hidden="false" class="notTouch" type="button" style="transform: scale(1);">
    Control
  </button>
</vime-control>
`;
